### PR TITLE
Update to 6.4-20240210

### PR DESCRIPTION
**Pull request by schribl**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

This is a try to use a newer patch level of ncurses to include potential bug and security fixes (similar to most distributions).

### CODE REVIEW

**Comment by schribl** [comment]
@conda-forge-admin, please rerender

**Comment by ekomarova** [comment]
@ocefpaf Could you please help to review this

File: recipe/meta.yaml
Comment (ocefpaf) [inline recipe/meta.yaml:10]: Sorry for my ignorance but source changes on important libraries are a bit scary. Do you have any background info on what/how/why we need to change the source here?

Comment (ekomarova) [reply recipe/meta.yaml:10]: The main problem is that the original sources https://invisible-island.net/ncurses/ncurses.faq.html#latest_version have not been supported for a long time and the last release was in 2022, and currently only patches are being released https://invisible-island.net/archives/ncurses/6.4/. But there are alternative sources, such as https://github.com/ThomasDickey/ncurses-snapshots, which are supported (ThomasDickey is the maintainer of ncurses) and which can be used to build the component without vulnerabilities. After all, version 6.4_20221231 contains quite a lot of CVE that have been fixed in latest versions. Speaking of these sources (http://invisible-mirror.net/archives/ncurses/current), I don't know much about them, but maybe @chribl can tell us why he suggests using exactly them

Comment (ocefpaf) [reply recipe/meta.yaml:10]: Thanks! That helps a lot!! @isuruf and/or @h-vetinari are usually more up-to-date than myself with these to judge how urgent we need to merge this update. I'll merge this in a few days if they don't comment until then.

Comment (jakirkham) [reply recipe/meta.yaml:10]: Could we please use this instead?

```suggestion
  url: https://ftp.gnu.org/gnu/ncurses/ncurses-{{ version }}.tar.gz
```

Comment (h-vetinari) [reply recipe/meta.yaml:10]: Generally, CVEs create a lot of sad (or angry) faces in environments where automated tooling enforces their absence. It's IMO in the interest of conda-forge to fix these things quickly.

That said, I don't know this specifc case, though mainly it appears we just need to use up-to-date sources. Pointing towards the upstream canonical repo is usually good enough for the bot to pick up updates. Fine by me to use the gnu server, or an up-to-date mirror on GitHub/gitlab (if there is one; not 100% sure our bot detects updates on the FTP server)

Comment (jmarshall) [reply recipe/meta.yaml:10]: I don't know this specific case either. However:

* The PR is to use the 6.4-20240210 patch snapshot rather than the 6.4 release (which is from late 2022).
* ftp.gnu.org does not contain the patch snapshot — the most recent tarball there is the 6.4 release, which is already the existing conda-forge package.
* The ThomasDickey/ncurses-snapshots GitHub repository almost certainly contains the same code as invisible-mirror.net, which is Thomas Dickey's web site. Indeed `https://www.gnu.org/software/ncurses/` redirects to invisible-island.net, which apparently invisible-mirror.net mirrors (and appears to also be controlled by TD).

Comment (schribl) [reply recipe/meta.yaml:10]: Sorry for causing so much confusion, I should have added more description in the beginning.. 
@jmarshall Already did the anaylsis in his comment, but for completeness I add my view when creating this PR (that was mostly about discussing the possibility of an ncurses update).

* ncurses does not have a official release since a while and an open CVE in the last release
* Thomas Dickey is releasing the current development snapshots after patches are applied (e.g the lastest from Sunday https://lists.gnu.org/archive/html/bug-ncurses/2024-03/msg00002.html )
* Also Linux distributions did update to user later ncurses versions to fix the CVE ( see e.g. Arch Linuxs patch here https://gitlab.archlinux.org/archlinux/packaging/packages/ncurses/-/commit/3c2606603aa4a5a3b2d29e560a1bc14986153f49 )
* Also from the GNU mailing list for ncurses and the mail from Thomas Dickey the invisible-island.net and invisible-mirror.net are official pages from Thomas Dickey AFAIK I was using the similar link to the Arch recipe

Open questions from my side about this PR:
* Is it acceptable to use a "beta" package for ncurses to mitigate the open CVE in nucrses?
* If so what is the best "beta" version to pick? (I just used the latest as example) Another approach could be e.g. the same version as Arch Linux as it is apparently already tested there
* When checking this I noticed that the links used in this PR are not online very long (that was also the reason to use the latest), e.g. right now the last patch release available on this link is from October 2023. So if we do use this approach at some point the recipe might be broken as the URL will be offline (approx. 5-6 month after the version was published)

So for the last point some other way could be to use the github repository @jmarshall  mentioned. If this is a legit source this might be the most flexible and future proof solution IMO.

Sorry again for the confusion and I am happy to do any recommended changes to this PR, the main purpose was to fix the CVE and start the communication, I am also pretty new to the conda-forge environment and did not knew where to best post this as according to this comment https://github.com/conda-forge/ncurses-feedstock/pull/71#issuecomment-1579262566 there are not many active maintainers of this recipe.

Comment (ekomarova) [reply recipe/meta.yaml:10]: Then my suggestion is to use these sources https://github.com/ThomasDickey/ncurses-snapshots by referring to the latest available tag, this will be the safest way

Comment (schribl) [reply recipe/meta.yaml]: > Then my suggestion is to use these sources https://github.com/ThomasDickey/ncurses-snapshots by referring to the latest available tag, this will be the safest way

I added a commit to change the recipe to retrieve the source from this GitHub mirror. Also I changed it a bit that it can handle both "full" versions and patched versions from GitHub and as a package version.

Comment (h-vetinari) [reply recipe/meta.yaml]: For anyone wondering, Thomas Dickey is the principal author of ncurses for the last ~25 years, see this long historical [overview](https://invisible-island.net/ncurses/ncurses.faq.html). His snapshots are about as official as it gets, certainly more than the GNU sources. They're also [updated](https://github.com/ThomasDickey/ncurses-snapshots/tags) very regularly, basically every week.

File: recipe/meta.yaml
Comment (ekomarova) [inline recipe/meta.yaml:3]: I also have a question about the logic that updates component versions, will it be able to update with the patch version? @ocefpaf @h-vetinari @jakirkham @jmarshall Do you know anything about how it works on conda-forge side? Perhaps the current logic will not be able to handle the modification of the version if it is such

Comment (ocefpaf) [reply recipe/meta.yaml:3]: We should try to use release for the auto-updating bot to work properly. I don't think it will work with a git tag.

File: recipe/meta.yaml
Comment (isuruf) [reply recipe/meta.yaml]: What does Debian do? Do they have patches that we can apply?

File: recipe/meta.yaml
Comment (ekomarova) [reply recipe/meta.yaml:3]: Well, they say about https://github.com/ThomasDickey/ncurses-snapshots.git, there are only tags here ☹️

Comment (ocefpaf) [reply recipe/meta.yaml:3]: You can still use the URL form, like:

```yaml
url: https://github.com/ThomasDickey/ncurses-snapshots/archive/refs/tags/{{ version }}.tar.gz
sha256: ....

File: recipe/meta.yaml
Comment (ekomarova) [reply recipe/meta.yaml]: It looks like they're still using 6.4_20221231 for the stable debian  version (although I'm not sure). One of the latest versions of ncurses in debian was 6.4_20240113-1, but available for Sid

File: recipe/meta.yaml
Comment (h-vetinari) [reply recipe/meta.yaml:3]: It's better to do any computation on the tag after it has been been picked up by the bot; otherwise, if you hardcode the patch value here, the bot will never find an update by itself. You can check the rstudio feedstock for an example of what I'm talking about.

**Comment by schribl** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml
Comment (schribl) [reply recipe/meta.yaml]: I did update it to my best knowledge based on @h-vetinari  and @ocefpaf  recommendations. I was trying to setup the autotick bot locally to be able to test if it was working, but I failed with understanding how (and if) it can be used to test if the recipe supports automatic updates.

**Comment by ekomarova** [comment]
@ocefpaf @fhoehle @jakirkham @mingwandroid Could you please review this again

**Review by ocefpaf — APPROVED** [review]
I'd like @h-vetinari +1 here too before we merge but, in case he is too busy to see this, I'll merge in a day or two.

File: recipe/meta.yaml
Comment (schribl) [reply recipe/meta.yaml:1]: This was intentional as `6.4` is not the same as `6.4.20240210`. Otherwise we might have updated patch level without a clear indication what version it is.

**Comment by ekomarova** [comment]
@ocefpaf This is just a question related to how conda-forge pipelines work and how the built bits are scanned before uploading to the channel. Is there any kind of validation and virus checking, for example with mcafee or smth?

**Pull request merged**
